climpred/bootstrap.py

```diff
@@ -173,81 +212,131 @@ def bootstrap_perfect_model(ds,
-        init skill."""
-        pv = ((simple_fct - init) > 0).sum('bootstrap') / init.bootstrap.size
-        if metric not in ['pearson_r', 'ppp']:  # positively oriented metrics
-            pv = 1 - pv
-        return pv
-
-    if compute_uninitialized_skill:
-        p_uninit_over_init = _pvalue_from_distributions(uninit, init)
-    else:
-        p_uninit_over_init, uninit_ci = None, None
-
-    if compute_persistence_skill:
-        p_pers_over_init = _pvalue_from_distributions(pers, init)
-
-    else:
-        p_pers_over_init, pers_ci = None, None
+    uninit = xr.concat(uninit, dim='bootstrap')
+    pers = xr.concat(pers, dim='bootstrap')
+
+    init_ci = _distribution_to_ci(init, ci_low, ci_high)
+    uninit_ci = _distribution_to_ci(uninit, ci_low, ci_high)
+    pers_ci = _distribution_to_ci(pers, ci_low_pers, ci_high_pers)
+
+    p_uninit_over_init = _pvalue_from_distributions(uninit, init)
+    p_pers_over_init = _pvalue_from_distributions(pers, init)
 
     # calc skill
     init_skill = compute_perfect_model(
-        ds, control, metric=metric, comparison=comparison, running=running,
-        reference_period=reference_period)
+        ds, metric=metric, comparison=comparison)
+    uninit_skill = uninit.mean('bootstrap')
+    pers_skill = compute_persistence(
+        ds, control, metric=metric)
+
+    # somehow there may be a member dim, which lets concat crash
+    if 'member' in init_skill:
+        del init_skill['member']
+    # wrap results together in one dataarray
+    skill = xr.concat([init_skill.squeeze(), uninit_skill, pers_skill], 'i')
+    skill['i'] = ['init', 'uninit', 'pers']
+
+    # probability that i beats init
+    p = xr.concat([p_uninit_over_init, p_pers_over_init], 'i')
+    p['i'] = ['uninit', 'pers']
+
+    # ci for each skill
+    ci = xr.concat([init_ci, uninit_ci, pers_ci],
+                   'i').rename({'quantile': 'results'})
+    ci['i'] = ['init', 'uninit', 'pers']
+
+    results = xr.concat([skill, p], 'results')
+    results['results'] = ['skill', 'p']
+    # (RXB) Drop any spurious coordinates that came along with results due to
+    # input ds. You can't concatenate if results and ci don't exactly match
+    # in coordinates. Maybe we can create a decorator to check this before
+    # anytime we merge.
+    if set(results.coords) != set(ci.coords):
+        res_drop = [c for c in results.coords if c not in ci.coords]
+        ci_drop = [c for c in ci.coords if c not in results.coords]
+        results = results.drop(res_drop)
+        ci = ci.drop(ci_drop)
+    results = xr.concat([results, ci], 'results')
+    return results
+
+
+def bootstrap_hindcast(dple,
+                       lens,
+                       reference,
+                       metric='pearson_r',
+                       comparison='e2r',
+                       bootstrap=5,
+                       sig=95,
+                       pers_sig=None):
+    """See bootstrap_perfect_model but for hindcasts."""
```

Comment:
Would be good to explicitly copy it over here in case someone runs `help(bootstrap_hindcast)`

Reply:
is there a way to automize this easily? I can just copy for sure.

Reply:
also for `classes`, all the changes seem tedious copy pasting for maintain. isnt there a clever way to automate this?

Reply:
I'm not aware of any way to automate the docstrings. And do you mean because classes are re-producing the docstrings of the functions they're calling? There might be some examples out there on how to deal with this.

Reply:
Maybe something like this https://docrep.readthedocs.io/en/latest/